File: app_pojavlauncher/src/main/java/net/kdt/pojavlaunch/prefs/screens/LauncherPreferenceMiscellaneousFragment.java
package net.kdt.pojavlaunch.prefs.screens;

import android.Manifest;
import android.app.Activity;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AlertDialog;
import androidx.preference.Preference;

import git.artdeell.mojo.R;

import net.kdt.pojavlaunch.LauncherActivity;
import net.kdt.pojavlaunch.PojavApplication;
import net.kdt.pojavlaunch.Tools;
import net.kdt.pojavlaunch.progresskeeper.ProgressKeeper;
import net.kdt.pojavlaunch.tasks.DataMigrator;
import net.kdt.pojavlaunch.utils.GLInfoUtils;
import net.kdt.pojavlaunch.utils.RendererCompatUtil;

import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.IOException;

public class LauncherPreferenceMiscellaneousFragment extends LauncherPreferenceFragment {

    private final ActivityResultLauncher<Uri> mMigrateLauncher = registerForActivityResult(
            new ActivityResultContracts.OpenDocumentTree(), (uri) -> {
                if(uri != null) {
                    new AlertDialog.Builder(getLauncherActivity())
                            .setTitle(R.string.migration_progress_warning_title)
                            .setMessage(R.string.migration_progress_warning_summary)
                            .setPositiveButton(android.R.string.ok, (d, w) -> new DataMigrator(getLauncherActivity(), uri).migrateData())
                            .setNegativeButton(android.R.string.cancel, null)
                            .show();
                }
            }
    );

    @Override
    public void onCreatePreferences(Bundle b, String str) {
        mVisibilityUpdater = this::updateVisibility;
        addPreferencesFromResource(R.xml.pref_misc);
        Preference driverPreference = requirePreference("zinkPreferSystemDriver");
        PackageManager packageManager = driverPreference.getContext().getPackageManager();
        boolean supportsTurnip = RendererCompatUtil.checkVulkanSupport(packageManager) && GLInfoUtils.getGlInfo().isAdreno();
        driverPreference.setVisible(supportsTurnip);
        Preference importPreference = requirePreference("runDataMigration");
        importPreference.setOnPreferenceClickListener(preference -> {
            if(ProgressKeeper.getTaskCount() > 0) {
                Toast.makeText(getContext(), R.string.tasks_ongoing, Toast.LENGTH_SHORT).show();
                return true;
            }
            mMigrateLauncher.launch(null);
            return true;
        });
        setupCacheClearPreference();
        setupMicrophoneRequestPreference();
        setupVintageStoryDeletePreference();
        updateVisibility();
    }

    private void setupVintageStoryDeletePreference() {
        Preference deletePreference = requirePreference("deleteVintageStory");
        deletePreference.setOnPreferenceClickListener(preference -> {
            if(!net.kdt.pojavlaunch.utils.jre.DotnetAssets.isInstalled()) {
                Toast.makeText(getContext(), R.string.preference_vintagestory_delete_not_installed, Toast.LENGTH_SHORT).show();
                return true;
            }
            new AlertDialog.Builder(getLauncherActivity())
                    .setTitle(R.string.preference_vintagestory_delete_title)
                    .setMessage(R.string.preference_vintagestory_delete_confirm)
                    .setPositiveButton(android.R.string.ok, (d, w) -> {
                        PojavApplication.sExecutorService.submit(() -> {
                            try {
                                net.kdt.pojavlaunch.utils.jre.DotnetAssets.removeAll();
                            } catch (IOException e) {
                                Tools.showErrorRemote(getLauncherActivity(), R.string.preference_vintagestory_delete_fail, e);
                                return;
                            }
                            Tools.runOnUiThread(() -> Toast.makeText(getLauncherActivity(), R.string.preference_vintagestory_delete_done, Toast.LENGTH_LONG).show());
                        });
                    })
                    .setNegativeButton(android.R.string.cancel, null)
                    .show();
            return true;
        });
    }

    private void updateVisibility(){
        requirePreference("microphoneAccessRequest").setVisible(!getLauncherActivity().checkForPermissionRationale(33, Manifest.permission.RECORD_AUDIO));
        requirePreference("clearMetadataCache").setVisible(new File(Tools.DIR_CACHE, "string_cache").exists());
    }

    @Override
    public void onResume() {
        super.onResume();
    }

    private void setupMicrophoneRequestPreference() {
        Preference mRequestMicrophonePermissionPreference = requirePreference("microphoneAccessRequest");
        Activity activity = getActivity();
        if(activity instanceof LauncherActivity) {
            mRequestMicrophonePermissionPreference.setOnPreferenceClickListener(preference -> {
                ((LauncherActivity) activity).askForPermission(23, Manifest.permission.RECORD_AUDIO);
                return true;
            });
        } else {
            mRequestMicrophonePermissionPreference.setVisible(false);
        }
    }
    private void setupCacheClearPreference() {
        Preference clearPreference = requirePreference("clearMetadataCache");
        clearPreference.setOnPreferenceClickListener(preference -> {
            if(ProgressKeeper.getTaskCount() > 0) {
                Toast.makeText(getContext(), R.string.tasks_ongoing, Toast.LENGTH_SHORT).show();
                return true;
            }
            PojavApplication.sExecutorService.submit(() -> {
                try {
                    FileUtils.deleteDirectory(new File(Tools.DIR_CACHE, "string_cache"));
                } catch (IOException e) {
                    Tools.showErrorRemote(getLauncherActivity(), R.string.preference_metadata_clear_fail, e);
                    return;
                }
                Tools.runOnUiThread(() -> {
                    Toast.makeText(getLauncherActivity(), R.string.preference_metadata_clear_complete, Toast.LENGTH_LONG).show();
                    updateVisibility();
                });
            });
            return true;
        });
    }
}

File: app_pojavlauncher/src/main/java/net/kdt/pojavlaunch/prefs/screens/LauncherPreferenceMiscellaneousFragment.kt
package net.kdt.pojavlaunch.prefs.screens

import android.Manifest
import android.app.Activity
import android.net.Uri
import android.os.Bundle
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.preference.Preference
import git.artdeell.mojo.R
import net.kdt.pojavlaunch.tasks.DataMigrator
import net.kdt.pojavlaunch.utils.GLInfoUtils
import net.kdt.pojavlaunch.utils.RendererCompatUtil

class LauncherPreferenceMiscellaneousFragment : LauncherPreferenceFragment() {

    private val mMigrateLauncher: ActivityResultLauncher<Uri?> = registerForActivityResult(
        ActivityResultContracts.OpenDocumentTree()
    ) { uri ->
        if (uri != null) {
            AlertDialog.Builder(getLauncherActivity())
                .setTitle(R.string.migration_progress_warning_title)
                .setMessage(R.string.migration_progress_warning_summary)
                .setPositiveButton(android.R.string.ok) { _, _ -> DataMigrator(getLauncherActivity(), uri).migrateData() }
                .setNegativeButton(android.R.string.cancel, null)
                .show()
        }
    }

    override fun onCreatePreferences(b: Bundle?, str: String?) {
        mVisibilityUpdater = Runnable { updateVisibility() }
        addPreferencesFromResource(R.xml.pref_misc)
        val driverPreference = requirePreference("zinkPreferSystemDriver")
        val packageManager = driverPreference.context.packageManager
        val supportsTurnip = RendererCompatUtil.checkVulkanSupport(packageManager) && GLInfoUtils.getGlInfo().isAdreno()
        driverPreference.isVisible = supportsTurnip
        val importPreference = requirePreference("runDataMigration")
        importPreference.setOnPreferenceClickListener {
            if (getLauncherActivity().hasActiveProcesses()) {
                Toast.makeText(context, R.string.tasks_ongoing, Toast.LENGTH_SHORT).show()
                return@setOnPreferenceClickListener true
            }
            mMigrateLauncher.launch(null)
            true
        }
        setupMicrophoneRequestPreference()
    }

    private fun updateVisibility() {
        requirePreference("microphoneAccessRequest").isVisible = !getLauncherActivity().checkForPermissionRationale(33, Manifest.permission.RECORD_AUDIO)
    }

    override fun onResume() {
        super.onResume()
    }

    private fun setupMicrophoneRequestPreference() {
        val mRequestMicrophonePermissionPreference = requirePreference("microphoneAccessRequest")
        val activity = activity
        if (activity is net.kdt.pojavlaunch.LauncherActivity) {
            mRequestMicrophonePermissionPreference.setOnPreferenceClickListener {
                activity.askForPermission(23, Manifest.permission.RECORD_AUDIO)
                true
            }
        } else {
            mRequestMicrophonePermissionPreference.isVisible = false
        }
        updateVisibility()
    }
}
